cachi2/core/package_managers/gomod.py

```diff
@@ -79,10 +79,146 @@ class GoPackage(_GolangModel):
 
     import_path: str
     standard: bool = False
-    module: Optional[GoModule]
+    module: Optional[ParsedModule]
     deps: list[str] = []
 
 
+class Module(NamedTuple):
+    """A Go module with relevant data for the SBOM generation"""
+    name: str
+    # name of the module before it was replaced
+    original_name: str
+    version: str
+    # The real URL to reach the module, differs from name in case of local replacements
+    url: str
+    main: bool = False
+
+    @property
+    def purl(self) -> str:
+        return f"pkg:golang/{self.url}@{self.version}?type=module"
+
+    def to_component(self) -> Component:
+        return Component(name=self.name, version=self.version, purl=self.purl)
+
+
+class Package(NamedTuple):
+    """A Go package with relevant data for the SBOM generation"""
+    name: str
+    module: Optional[Module]
+    standard: bool = False
+
+    @property
+    def relative_path(self) -> str:
+        return self.name.replace(self.module.name, "")
+
+    @property
+    def purl(self) -> str:
+        if self.standard:
+            return f"pkg:golang/{self.name}"
+
+        return f"pkg:golang/{self.module.url}{self.relative_path}@{self.module.version}?type=package"
+
+    def to_component(self) -> Component:
+        if self.standard:
+            version = None
+        else:
+            version = self.module.version
+
+        return Component(name=self.name, version=version, purl=self.purl)
+
+
+class ModuleFactory(NamedTuple):
+    main_module: Module
+    app_dir: RootedPath
+
+    def from_parsed_module(self, module: ParsedModule) -> "Module":
+        if not (replace := module.replace):
+            name = module.path
+            version = module.version
+            original_name = name
+            url = name
+        elif replace.version:
+            # module/name v1.0.0 => replace/name v1.2.3
+            name = replace.path
+            version = replace.version
+            original_name = module.path
+            url = name
+        else:
+            # module/name v1.0.0 => ./local/path
+            name = module.path
+            module_path = self.app_dir.join_within_root(module.replace.path)
+            version = _get_golang_version(module.path, module_path)
+            url = self._resolve_url_for_local_replacement(module)
+            original_name = name
+
+        return Module(name=name, version=version, original_name=original_name, url=url)
+
+    def _resolve_url_for_local_replacement(self, module: ParsedModule):
```

Comment:
There's probably a better way to do this. Pathlib was not helpful because we're using URLs here, not actual file paths.

Reply:
ah, but you want a subpath from the module dir, not from the repo root :thinking:

Reply:
Otherwise, I think this function does roughly the same thing as [normpath](https://docs.python.org/3/library/os.path.html#os.path.normpath) (which, as the docs say, may change the semantic meaning of a path)

Reply:
I improved the function a bit. I'd love to replace it with normpath, but there's the thing about the reverted slashes on Windows. I'm very sure Windows is not a scenario that we will be running into, but we probably should at least throw an exception to help the future generations.

Wdyt?

Reply:
You can un-revert the slashes with `Path.as_posix()`:

```python
normalized_but_platform_specific_path = os.path.normpath(p)
normalized_posix_path = Path(normalized_but_platform_specific_path).as_posix()
```